{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ffee2814",
   "metadata": {},
   "source": [
    "### Import libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "620820b2",
   "metadata": {},
   "source": [
    "pandas - data analysis and manipulation\n",
    "\n",
    "numpy - support for multidimensional arrays and matrices\n",
    "\n",
    "genomepy -  handling genomes and gene annotations\n",
    "\n",
    "matplotlib - creating visualisations \n",
    "\n",
    "SciPy - scientific and technical computing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a492bba8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: genomepy in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (0.14.0)\n",
      "Requirement already satisfied: loguru in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (0.6.0)\n",
      "Requirement already satisfied: mygene in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (3.2.2)\n",
      "Requirement already satisfied: filelock in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (3.6.0)\n",
      "Requirement already satisfied: mysql-connector-python in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (8.0.32)\n",
      "Requirement already satisfied: numpy in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (1.21.5)\n",
      "Requirement already satisfied: appdirs in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (1.4.4)\n",
      "Requirement already satisfied: pandas in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (1.4.4)\n",
      "Requirement already satisfied: requests in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (2.28.1)\n",
      "Requirement already satisfied: pyfaidx>=0.5.7 in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (0.7.1)\n",
      "Requirement already satisfied: click in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (8.0.4)\n",
      "Requirement already satisfied: biopython>=1.73 in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (1.80)\n",
      "Requirement already satisfied: norns>=0.1.5 in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (0.1.6)\n",
      "Requirement already satisfied: colorama in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (0.4.5)\n",
      "Requirement already satisfied: tqdm>=4.51 in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (4.64.1)\n",
      "Requirement already satisfied: diskcache in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from genomepy) (5.4.0)\n",
      "Requirement already satisfied: nose in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from norns>=0.1.5->genomepy) (1.3.7)\n",
      "Requirement already satisfied: pyyaml in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from norns>=0.1.5->genomepy) (6.0)\n",
      "Requirement already satisfied: setuptools>=0.7 in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from pyfaidx>=0.5.7->genomepy) (63.4.1)\n",
      "Requirement already satisfied: six in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from pyfaidx>=0.5.7->genomepy) (1.16.0)\n",
      "Requirement already satisfied: biothings-client>=0.2.6 in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from mygene->genomepy) (0.2.6)\n",
      "Requirement already satisfied: protobuf<=3.20.3,>=3.11.0 in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from mysql-connector-python->genomepy) (3.20.3)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from pandas->genomepy) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from pandas->genomepy) (2022.1)\n",
      "Requirement already satisfied: charset-normalizer<3,>=2 in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from requests->genomepy) (2.0.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from requests->genomepy) (2022.9.24)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from requests->genomepy) (1.26.11)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/marcinmaniak/opt/anaconda3/lib/python3.9/site-packages (from requests->genomepy) (3.3)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install genomepy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f8eb06cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import genomepy\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59df998b",
   "metadata": {},
   "source": [
    "### Download genome assembly with genomepy and read in the gene annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f5090fcf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSMUSG00000102693.2;gene_id=ENSMUSG0000010...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSMUST00000193812.2;Parent=ENSMUSG00000102...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENSMUST00000193812.2:1;Parent=ENSMUST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSMUSG00000064842.3;gene_id=ENSMUSG0000006...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSMUST00000082908.3;Parent=ENSMUSG00000064...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname   source     feature    start      end score strand frame  \\\n",
       "0    chr1   HAVANA        gene  3143476  3144545     .      +     .   \n",
       "1    chr1   HAVANA  transcript  3143476  3144545     .      +     .   \n",
       "2    chr1   HAVANA        exon  3143476  3144545     .      +     .   \n",
       "3    chr1  ENSEMBL        gene  3172239  3172348     .      +     .   \n",
       "4    chr1  ENSEMBL  transcript  3172239  3172348     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  ID=ENSMUSG00000102693.2;gene_id=ENSMUSG0000010...  \n",
       "1  ID=ENSMUST00000193812.2;Parent=ENSMUSG00000102...  \n",
       "2  ID=exon:ENSMUST00000193812.2:1;Parent=ENSMUST0...  \n",
       "3  ID=ENSMUSG00000064842.3;gene_id=ENSMUSG0000006...  \n",
       "4  ID=ENSMUST00000082908.3;Parent=ENSMUSG00000064...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomepy.install_genome(\"GRCm39\", \"NCBI\", annotation = False)\n",
    "\n",
    "gencode = pd.read_table(\"gencode.vM32.annotation.gff3.gz\", comment=\"#\", sep = \"\\t\", names = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute'])\n",
    "gencode.head() "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5684c1e1",
   "metadata": {},
   "source": [
    "### Select genes only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e429bf69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=ENSMUSG00000102693.2;gene_id=ENSMUSG0000010...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=ENSMUSG00000064842.3;gene_id=ENSMUSG0000006...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3276124</td>\n",
       "      <td>3741721</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=ENSMUSG00000051951.6;gene_id=ENSMUSG0000005...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3322980</td>\n",
       "      <td>3323459</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=ENSMUSG00000102851.2;gene_id=ENSMUSG0000010...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3435954</td>\n",
       "      <td>3438772</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=ENSMUSG00000103377.2;gene_id=ENSMUSG0000010...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname   source feature    start      end strand  \\\n",
       "0    chr1   HAVANA    gene  3143476  3144545      +   \n",
       "1    chr1  ENSEMBL    gene  3172239  3172348      +   \n",
       "2    chr1   HAVANA    gene  3276124  3741721      -   \n",
       "3    chr1   HAVANA    gene  3322980  3323459      +   \n",
       "4    chr1   HAVANA    gene  3435954  3438772      -   \n",
       "\n",
       "                                           attribute  \n",
       "0  ID=ENSMUSG00000102693.2;gene_id=ENSMUSG0000010...  \n",
       "1  ID=ENSMUSG00000064842.3;gene_id=ENSMUSG0000006...  \n",
       "2  ID=ENSMUSG00000051951.6;gene_id=ENSMUSG0000005...  \n",
       "3  ID=ENSMUSG00000102851.2;gene_id=ENSMUSG0000010...  \n",
       "4  ID=ENSMUSG00000103377.2;gene_id=ENSMUSG0000010...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_genes = gencode[(gencode.feature == \"gene\")][['seqname', 'source', 'feature','start', 'end', 'strand', 'attribute']].copy().reset_index().drop('index', axis=1)\n",
    "gencode_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f37b49f",
   "metadata": {},
   "source": [
    "### Extract gene names and gene types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dc325f8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_info(x):\n",
    "    g_name = list(filter(lambda x: 'gene_name' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_type = list(filter(lambda x: 'gene_type' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    return (g_name, g_type)\n",
    "gencode_genes[\"gene_name\"], gencode_genes[\"gene_type\"] = zip(*gencode_genes.attribute.apply(lambda x: gene_info(x)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42aea026",
   "metadata": {},
   "source": [
    "### Select protein-coding genes only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e45768e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3276124</td>\n",
       "      <td>3741721</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=ENSMUSG00000051951.6;gene_id=ENSMUSG0000005...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4069780</td>\n",
       "      <td>4479464</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=ENSMUSG00000025900.14;gene_id=ENSMUSG000000...</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4561154</td>\n",
       "      <td>4567577</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=ENSMUSG00000025902.14;gene_id=ENSMUSG000000...</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4843429</td>\n",
       "      <td>4855962</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=ENSMUSG00000033845.14;gene_id=ENSMUSG000000...</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4878011</td>\n",
       "      <td>4918633</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=ENSMUSG00000025903.15;gene_id=ENSMUSG000000...</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source feature    start      end strand  \\\n",
       "0    chr1  HAVANA    gene  3276124  3741721      -   \n",
       "1    chr1  HAVANA    gene  4069780  4479464      -   \n",
       "2    chr1  HAVANA    gene  4561154  4567577      -   \n",
       "3    chr1  HAVANA    gene  4843429  4855962      -   \n",
       "4    chr1  HAVANA    gene  4878011  4918633      +   \n",
       "\n",
       "                                           attribute gene_name       gene_type  \n",
       "0  ID=ENSMUSG00000051951.6;gene_id=ENSMUSG0000005...      Xkr4  protein_coding  \n",
       "1  ID=ENSMUSG00000025900.14;gene_id=ENSMUSG000000...       Rp1  protein_coding  \n",
       "2  ID=ENSMUSG00000025902.14;gene_id=ENSMUSG000000...     Sox17  protein_coding  \n",
       "3  ID=ENSMUSG00000033845.14;gene_id=ENSMUSG000000...    Mrpl15  protein_coding  \n",
       "4  ID=ENSMUSG00000025903.15;gene_id=ENSMUSG000000...    Lypla1  protein_coding  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_genes = gencode_genes[gencode_genes['gene_type'] == 'protein_coding'].reset_index().drop('index', axis=1)\n",
    "gencode_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23b63e19",
   "metadata": {},
   "source": [
    "### Index by gene name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5b328615",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3276124</td>\n",
       "      <td>3741721</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=ENSMUSG00000051951.6;gene_id=ENSMUSG0000005...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rp1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4069780</td>\n",
       "      <td>4479464</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=ENSMUSG00000025900.14;gene_id=ENSMUSG000000...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sox17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4561154</td>\n",
       "      <td>4567577</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=ENSMUSG00000025902.14;gene_id=ENSMUSG000000...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mrpl15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4843429</td>\n",
       "      <td>4855962</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=ENSMUSG00000033845.14;gene_id=ENSMUSG000000...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lypla1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4878011</td>\n",
       "      <td>4918633</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=ENSMUSG00000025903.15;gene_id=ENSMUSG000000...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          seqname  source feature    start      end strand  \\\n",
       "gene_name                                                    \n",
       "Xkr4         chr1  HAVANA    gene  3276124  3741721      -   \n",
       "Rp1          chr1  HAVANA    gene  4069780  4479464      -   \n",
       "Sox17        chr1  HAVANA    gene  4561154  4567577      -   \n",
       "Mrpl15       chr1  HAVANA    gene  4843429  4855962      -   \n",
       "Lypla1       chr1  HAVANA    gene  4878011  4918633      +   \n",
       "\n",
       "                                                   attribute       gene_type  \n",
       "gene_name                                                                     \n",
       "Xkr4       ID=ENSMUSG00000051951.6;gene_id=ENSMUSG0000005...  protein_coding  \n",
       "Rp1        ID=ENSMUSG00000025900.14;gene_id=ENSMUSG000000...  protein_coding  \n",
       "Sox17      ID=ENSMUSG00000025902.14;gene_id=ENSMUSG000000...  protein_coding  \n",
       "Mrpl15     ID=ENSMUSG00000033845.14;gene_id=ENSMUSG000000...  protein_coding  \n",
       "Lypla1     ID=ENSMUSG00000025903.15;gene_id=ENSMUSG000000...  protein_coding  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_genes = gencode_genes.set_index('gene_name')\n",
    "gencode_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13fec93c",
   "metadata": {},
   "source": [
    "### Create dictionary of genes with their esitmated elongation rates "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7c73d32d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Kdm2a': 1482.0664, 'Ankib1': 1992.554, 'Pid1': 1612.144, 'Wapl': 896.2004, 'Kdm2b': 1526.1148, 'Nol10': 1347.6652, 'Zfyve26': 595.1316, 'Septin11': 2407.198, 'Usp13': 1554.2792, 'Ercc6l2': 1017.0216, 'Snx13': 2688.8208, 'C2cd3': 1330.8956, 'N4bp2': 1256.5544, 'Crebbp': 2558.938, 'Otud7b': 1150.5932, 'Usp32': 1978.4756, 'Smg1': 2356.2876, 'Usp31': 2220.418, 'Cblb': 2476.694, 'Fancd2': 866.8708, 'Fam120a': 2233.7096, 'Spata13': 1694.3776, 'Ddx31': 1470.9436, 'Gucy1a2': 2400.0, 'Ralgapa2': 2535.3224, 'Gfod1': 1088.1276, 'Dock1': 2088.1196, 'Thoc2': 1880.0, 'Garem1': 2167.7232, 'Bclaf3': 1256.1996, 'Rfx7': 2487.55, 'Zyg11b': 884.404, 'Mef2a': 2258.272, 'Ak8': -440.0, 'Rabgap1': 1294.5556, 'Erich1': 1500.418, 'Slc38a6': 1457.3628, 'Fam117b': 1868.22, 'Cep350': 2055.0, 'Tcerg1': 1439.7512, 'Plekhm3': 730.2772, 'Pdxdc1': 1525.1096, 'Myo6': 2452.754, 'Edem3': 1871.2152, 'Mtmr7': 1052.718, 'Kntc1': 848.076, 'Apaf1': 674.9648, 'Raph1': 2149.4576, 'Fkbp15': 1331.1188, 'Med14': 1716.472, 'Pik3r1': 2535.0516, 'Usp6nl': 3063.13, 'Btaf1': 2318.9184, 'Rsbn1l': 2941.3584, 'Ccdc171': 1777.8568, 'Kmt2a': 1739.594, 'Tecpr2': 1448.7764, 'Bdp1': 1498.872, 'Gls': 1801.6036, 'Cep152': 1296.5076, 'Taf4': 2239.596, 'Stim2': 1655.6516, 'Arhgef17': 884.6836, 'Mtr': 1391.7192, 'Kif16b': 1926.252, 'Kat6a': 2975.696, 'Heatr5b': 1564.0776, 'Virma': 1290.3108, 'Eml5': 1591.1944, 'Arid4a': 1372.2092, 'Phip': 1893.6028, 'Ercc6': 1259.9872, 'Slf2': 884.68, 'Pbrm1': 1128.9184, 'Uggt2': 2089.784, 'Rsf1': 3329.3664, 'Ibtk': 1000.8292, 'Prrc2c': 1606.252, 'Ric1': 2872.912, 'Pds5a': 2192.4876, 'Myo5c': 1300.7832, 'Agl': 543.7144, 'Skic3': 1292.8752, 'Prdm2': 2075.722, 'Camsap2': 2220.26, 'Trrap': 1122.5628, 'Arhgap21': 3004.9132, 'Dmxl1': 2010.2772, 'Kidins220': 1459.008, 'Ankrd11': 2713.236, 'Amotl1': 780.192, 'Wdr70': 1500.2388, 'Phf20l1': 1831.4352, 'Ccdc15': 1137.5624, 'Ino80d': 1388.3536, 'Iqck': 6040.0, 'Brwd3': 1814.8176, 'Gnb1l': 1575.1532, 'Pphln1': 1816.8576, 'Rere': 2263.096, 'Arfgef2': 1464.456, 'Asap2': 3000.9064, 'Cep170': 1520.0, 'Taf4b': 1570.6428, 'Kazn': 1989.412, 'C2cd5': 917.3156, 'Acvr1': 2073.0636, 'Lnpk': 2108.8268, 'Ppm1h': 1373.8784, 'Itsn1': 901.2976, 'Snx12': 1153.0288, 'Rbfox2': 2526.7976, 'Reps1': 2294.6088, 'Ccdc6': 2406.252, 'Eif2b3': 2163.5968, 'Tbc1d9': 2045.244, 'Atosa': 1913.42, 'Gatad2a': 3127.782, 'Phf8': 1259.976, 'Rnf13': 1697.6916, 'Slain2': 1516.9964, 'Lima1': 1470.152, 'Mtcl1': 1760.0, 'Clasp2': 930.2772, 'Zmym4': 1497.0864, 'Arhgap17': 2287.4096, 'Kit': 1767.5536, 'Pde7a': 2027.6736, 'Lrrc8d': 789.7752, 'Pard3': 1933.0196, 'Gpbp1': 1075.0544, 'Qser1': 1298.434, 'Ndufaf2': 1381.2268, 'Rbm47': 316.54, 'Atp13a3': 2135.1772, 'Epb41': 1134.89, 'Oxr1': 1741.084, 'Raly': 2314.448, 'Tcf7l2': 2538.942, 'Fam13c': 774.4828, 'Gpd2': 2160.4984, 'Yeats2': 1747.6712, 'Zfp710': 1647.5536, 'Ptpn21': 1726.224, 'Gpatch8': 2349.7896, 'Pacsin2': 2497.7848, 'Macroh2a1': 1051.1388, 'Qki': 2516.3404, 'Adcy3': 970.7428, 'Fip1l1': 1966.5432, 'Pum1': 1607.5536, 'Slc44a1': 1869.79, 'Tasp1': 1070.296, 'Ccdc85c': 1800.0, 'Ralgps2': 1819.42, 'Pum2': 2322.864, 'Smg7': 2134.5556, 'Ptk2b': 1226.8388, 'Miga1': 1231.7488, 'Rab11fip3': 560.5784, 'Fnip2': 2821.872, 'Ythdc2': 648.0352, 'Mon2': 1042.656, 'Cog5': 1858.3288, 'Cmip': 2282.4536, 'Evl': 2080.0, 'Dlg5': 2397.3756, 'Tjp1': 2323.0664, 'Psma8': 2483.648, 'Rab3gap2': 2258.3876, 'Tank': 1125.5, 'Add3': 2955.2, 'Cert1': 1492.5896, 'Nrf1': 2519.042, 'Nt5c2': 2506.11, 'Tent4b': 2142.552, 'Slk': 1777.3628, 'Cyfip1': 1960.3176, 'Pals2': 2446.252, 'Lrrfip2': 2569.2404, 'Mia2': 1403.3476, 'Zfp638': 1433.1332, 'Sp4': 1885.1048, 'Dennd1b': 2460.8792, 'Cfap77': 880.0, 'Tbc1d31': 1591.6796, 'Wwtr1': 2490.2772, 'Cdk19': 2627.6392, 'Ctps2': 944.6124, 'Ppfibp1': 2553.2056, 'Depdc5': 1434.078, 'Wdr59': 1560.0, 'Mthfd1l': 1105.0076, 'Wdr33': 1800.0, 'Ahcyl2': 2100.4056, 'Prkd3': 2310.3992, '3110009E18Rik': 599.1008, 'Strn3': 2267.716, 'Rbm41]': 1497.0708, 'Gng12': 2520.0, 'Etl4': 1445.1048, 'Elavl2': 1437.162, 'Lclat1': 2703.0988, 'Dcun1d4': 1965.1048, 'Ppfia1': 1964.9848, 'Foxp1': 2865.8968, 'Itsn2': 2046.4232, 'Zmynd11': 1273.8216, 'Spag9': 2412.564, 'Ift43': 993.5084, 'Clec16a': 824.69, 'Map4': 2200.0, 'Golm2': 1459.488, 'Susd6': 2328.56, 'Acvr2a': 1944.5512, 'Adam10': 1957.2196, 'Apc': 2363.566, 'Blm': 1560.6464, 'Bmpr2': 2138.3356, 'Birc6': 1953.1244, 'Scarb2': 1624.6972, 'Chd1': 1532.1356, 'Cit': 2242.6628, 'Cpd': 2094.5556, 'Hivep1': 1957.1432, 'Elf1': 3389.5612, 'Mark2': 1736.46, 'Epc1': 4071.4976, 'Epha4': 2901.0776, 'Eps15': 2088.272, 'Smarcad1': 1785.804, 'Etv1': 2089.5976, 'Etv6': 2356.8404, 'Ezh2': 1771.9008, 'Fgf10': 2297.394, 'Gnaq': 2509.4984, 'Gnb1': 1839.47, 'Gpc4': 2136.1552, 'Grb2': 1892.2132, 'Plpp1': 1179.484, 'Hsd17b4': 1600.0, 'Hspg2': 2081.1344, 'Lrig1': 2214.8056, 'Itga6': 1853.9392, 'Itgav': 1587.576, 'Kif5c': 2297.012, 'Kpna3': 2566.5364, 'Lama1': 1828.092, 'Lrp1': 741.68, 'Lrp5': 1877.0112, 'Lrp6': 2210.556, 'Smad1': 1454.3796, 'Anapc1': 2019.826, 'Abcc1': 1583.7272, 'Mertk': 2240.0, 'Mgmt': 1548.036, 'Me1': 3384.0868, 'Map1b': 2016.022, 'Myo7a': 1520.012, 'Ncoa3': 2756.8896, 'Enah': 2747.4576, 'Neo1': 2994.384, 'Nfib': 2310.4312, 'Nfyc': 1821.9612, 'Nid2': 1288.2708, 'Ints6': 1666.25, 'Pcnt': 1016.5232, 'Pepd': 1564.0664, 'Prkci': 2098.02, 'Prkcz': 1839.0304, 'Pola1': 2343.1732, 'Map2k1': 2013.6256, 'Npepps': 2240.036, 'Ptdss1': 1799.056, 'Pten': 1642.4536, 'Ptpn14': 1320.0, 'Rock1': 2223.248, 'Rock2': 2309.43, 'Polr1a': 1046.252, 'Atxn2': 1936.8388, 'Exoc4': 2976.2164, 'Glg1': 2323.0664, 'Map2k4': 2432.1156, 'St3gal1': 3614.064, 'Clpb': 2204.6044, 'Smarcc1': 2016.7192, 'Sntb2': 1180.928, 'Swap70': 2200.0, 'Tcf7l1': 1490.8876, 'Tgfbr1': 2484.848, 'Tpd52': 2320.0, 'Tpp2': 1875.4156, 'Ttc3': 1624.9408, 'Kdm6a': 3272.4744, 'Uty': 1721.148, 'Atrx': 2347.9716, 'Xrcc5': 1000.0, 'Yes1': 2457.3628, 'Ankfy1': 1046.7664, 'Ap1g1': 2378.5092, 'Ap3b1': 2141.1988, 'Aplp2': 2337.3628, 'Aqr': 1358.8176, 'Arnt': 1998.3388, 'Atf2': 1706.2024, 'Atrn': 1648.2404, 'Brca1': 790.1808, 'Btrc': 1350.4356, 'Capzb': 1224.9736, 'Ctnna1': 1999.688, 'Runx1': 2088.524, 'Rb1cc1': 1520.0, 'Cd2ap': 2243.3476, 'Dpysl2': 1453.8528, 'E2f3': 1378.5536, 'Eif2ak3': 1823.2548, 'Ext2': 1687.6196, 'Fhl1': 671.36, 'Fkbp5': 1648.7948, 'Fn1': 1440.0, 'Fut9': 1521.4976, 'Nr6a1': 1817.8004, 'Gna12': 2036.4944, 'Gng2': 1771.8444, 'Htt': 2003.2948, 'Hira': 978.342, 'Igf1r': 2420.8956, 'Itpr1': 2164.3276, 'Kif15': 1200.0, 'Lamc1': 3403.2948, 'Lef1': 2183.7364, 'Mad1l1': 1159.9548, 'Smad2': 2342.1092, 'Man1a2': 3052.9444, 'Melk': 1647.8672, 'Mpdz': 2960.0, 'Msh3': 690.77, 'Msn': 1680.024, 'Nck2': 1426.8688, 'Ndufs4': 1678.1724, 'Nedd4': 2430.6636, 'Nf2': 2294.6088, 'Notch2': 2412.1356, 'Pikfyve': 1561.4272, 'Prkacb': 1296.7832, 'Pou2f1': 3099.9024, 'Prep': 1760.0, 'Prkdc': 839.72, 'Cyth3': 2972.096, 'Ptgfrn': 1800.1704, 'Ptpn12': 2191.8696, 'Rbl1': 667.562, 'Rfc1': 2052.2588, 'Rps6ka2': 1770.5104, 'Uimc1': 1438.5896, 'Ncor1': 1329.5968, 'Smarca4': 1298.1964, 'Sptlc2': 2015.68, 'Strn': 2207.044, 'Suclg2': 2020.0644, 'Tjp2': 2190.1192, 'Tle1': 2055.1712, 'Tle4': 2295.172, 'Ppp1r13b': 1401.1, 'Nr2c2': 2635.4596, 'Usp12': 1604.088, 'Zfp148': 2432.8156, 'Zfr': 2156.2016, 'Coro1c': 1955.0904, 'Cfdp1': 1166.754, 'Farsb': 1405.1048, 'Hs2st1': 2764.8684, 'Tlk2': 2726.4916, 'Xrn1': 1245.9896, 'Xrn2': 1315.9688, 'Cop1': 2259.3664, 'Fto': 2507.004, 'Map3k2': 2527.6168, 'Map3k3': 1739.372, 'Mapk1': 2339.6092, 'Orc5': 1309.0704, 'Plod2': 1886.4828, 'Homer2': 3800.0, 'Ppp2r5e': 2421.1476, 'Mprip': 2820.0352, 'Cul1': 1808.0144, 'Polk': 1393.658, 'Epb41l4a': 2614.46, 'Eif2ak1': 4800.0, 'Ryk': 2153.8356, 'Sema4d': 2423.6052, 'Eif2ak4': 1120.0, 'Trp53bp1': 1280.9168, 'Mkln1': 2056.3132, 'Galnt1': 2236.6516, 'Baz1a': 2133.0704, 'Ror2': 2323.748, 'Tpk1': 3321.9436, 'Zfp292': 2517.4072, 'Fbxw4': 2280.0, 'Sufu': 1657.3592, 'Greb1': 2109.9712, 'Atp11a': 1485.6308, 'Reck': 1048.0948, 'Gpc3': 630.9596, 'Mapk8': 2098.576, 'Cul3': 2494.5732, 'Smad3': 1793.0944, 'Ube2k': 1976.616, 'Atxn10': 3052.9424, 'Vti1a': 1160.0, 'Atp6v0a1': 853.3264, 'Nckap1': 1612.1504, 'Celf1': 2937.6916, 'Arhgef7': 1926.3956, 'Copg2': 2139.976, 'Akap7': 1198.2456, 'Cadm1': 962.4928, 'Pcnx': 2327.2672, 'Nup210': 1733.2488, 'Tmem131': 2381.054, 'Rabep1': 2560.0, 'Sfmbt1': 2923.1212, 'Uso1': 1905.3684, 'Ammecr1': 2927.2, 'Sh3gl2': 2071.5948, 'Ptbp2': 2207.6268, 'Rev1': 2152.4236, 'Tspan5': 3089.012, 'Wwox': 1955.6556, 'Hsd17b12': 2781.2108, 'Pias1': 2680.0, 'Diaph3': 2572.3928, 'Hbs1l': 1972.112, 'Pfkp': 1587.954, 'Sae1': 1332.1356, 'Snd1': 752.0204, 'Pex14': 2160.0, 'Aatf': 932.1376, 'Gsk3b': 2081.2596, 'Atr': 1034.228, 'Arih1': 2163.9004, 'Ranbp9': 2675.1796, 'Mbnl1': 2070.2032, 'Mtor': 847.0756, 'Trim44': 1566.364, 'Lmbr1': 1519.0784, 'Ank': 1829.4888, 'Xpo4': 2188.9004, 'Brd4': 2730.1524, 'Tbl1x': 1592.4236, 'Jph1': 514.1116, 'Jmy': 1853.2224, 'Rad18': 1807.2436, 'Tmeff1': 2208.6844, 'Fzd3': 1200.0, 'Dennd5a': 1547.564, 'Nectin3': 2091.1644, 'Ngly1': 520.0, 'Nup160': 1012.3596, 'Mark3': 1805.7236, 'Rab2a': 2062.688, 'Huwe1': 1241.284, 'Psmd14': 2151.6928, 'Rnf130': 2134.484, 'Pcbp3': 1911.7256, 'Nek7': 1577.9428, 'Nek6': 1403.164, 'Syt9': 1858.6908, 'Niban1': 705.0992, 'Cables1': 1029.3636, 'Ube4b': 1704.9072, 'Myh9': 2457.4508, 'Nup98': 1640.0, 'Xpo7': 2941.2268, 'Eefsec': 2691.2548, 'Slc30a7': 1738.0492, 'Agtpbp1': 1267.4516, 'Hacd2': 1475.9688, 'Pcm1': 1804.7576, 'Igf2bp3': 2087.804, 'Ern1': 1991.5244, 'Stt3b': 2129.1428, 'Rrbp1': 1641.8844, 'Desi2': 1432.4236, 'Spop': 1120.6632, 'Dtd1': 1950.87, 'Acer3': 1434.248, 'Smurf2': 2643.5244, 'Cenpp': 2003.228, 'Fam107b': 1559.992, 'Metap1d': 1170.8384, 'Pigk': 1889.9476, 'Dph6': 1209.536, 'Ctnnbl1': 2007.8984, 'Col27a1': 367.2508, 'Gapvd1': 1528.046, 'Dtnbp1': 1367.192, 'Hmg20a': 1472.1808, 'Lonp2': 994.148, 'Wwp2': 1363.3848, 'Bzw2': 1507.33, 'Thap4': 1022.9548, 'Ola1': 2255.68, 'Ube2g1': 2440.0, 'Dipk1a': 2743.24, 'Cwc27': 1892.9444, 'Dock7': 1849.2628, 'Zc3h13': 1857.4584, 'Herc4': 1547.4576, 'Lims1': 2336.7856, 'Ccdc50': 1960.8776, 'Ahi1': 600.0, 'Gstcd': 1991.9716, 'Ctdp1': 969.642, 'Alkbh8': 1160.0, 'Stx17': 1480.0, 'Pkp4': 2179.2208, 'Gpatch2': 1435.2612, 'Rnf115': 1645.816, 'Uxs1': 1780.174, 'Cmc1': 1285.7924, 'Klhl7': 1910.4936, 'Ccny': 2070.5584, 'Cspp1': 1877.0152, 'Ino80': 882.6772, 'Fam135a': 1800.0, 'Ift80': 1283.3656, 'Cep57l1': 409.7664, 'Agpat4': 1054.148, 'Zdhhc21': 1421.7476, 'Vps53': 1874.2072, 'Drosha': 1460.562, 'Svop': 428.6076, 'Zfp644': 2262.162, 'Ubac2': 2103.2024, 'Ubap2': 1012.8316, 'Anapc10': 1703.042, 'Stx18': 1663.1376, 'Kmt2e': 1864.5564, 'Ddah1': 1704.768, 'Ccser2': 1964.666, 'Cdk13': 1813.8056, 'Arhgef12': 2102.912, 'Pir': 1442.8364, 'Smyd3': 1779.9044, 'Rnf157': 806.4276, 'Dzip3': 2070.4176, 'Psmd1': 1059.648, 'Polr3b': 1158.7268, 'Mipep': 1089.3528, 'Bbx': 1533.4836, 'Nup205': 1232.8272, 'Cwf19l2': 1666.1416, 'Usp15': 1831.668, 'Zranb3': 1687.8368, 'Nbas': 1270.728, 'Ccdc57': 1049.012, 'Ccdc191': 1280.0, 'Ppp1r12a': 2791.9244, 'Itfg1': 1900.5764, 'Ppp2r2a': 2802.4324, 'Sulf2': 356.2248, 'Rexo5': 1234.942, 'Atxn7l1': 3416.4396, 'Slx4ip': 2658.3876, 'Lrpprc': 1162.0932, 'Rbm33': 2087.8696, 'Immp1l': 1295.028, 'Setd3': 1975.322, 'Pan3': 1792.076, 'Tmem135': 2400.0, 'Tdp1': 1007.0648, '3110082I17Rik': 2312.5704, 'Dcbld2': 2209.012, 'Smap1': 2431.868, 'Camkmt': 1041.522, 'Ttc7': 979.2092, 'Whrn': 515.6932, 'Herc3': 1342.7528, 'Als2': 1297.6916, 'Parn': 1320.3056, 'Rbm19': 579.4656, 'Armc8': 2520.0, 'Xpo6': 2603.8616, 'Aven': 2125.7236, 'Mtmr3': 1706.5304, 'Smchd1': 2095.472, 'Hsf2bp': 2232.696, 'Ppp6r3': 2252.6584, 'Sppl3': 2994.7256, 'Pik3cb': 2024.3588, 'Uhrf1bp1l': 2177.6596, 'Rbm6': 2179.9876, 'Shld2': 2660.2592, 'Dnaaf9': 1451.5872, 'Nln': 1451.0544, 'Zdhhc20': 1406.5432, 'Rab30': 1548.9528, 'Zmym2': 2052.1836, 'Erp44': 2717.1928, 'Lmf1': 742.852, 'Atg2b': 1445.1048, 'Snx7': 2208.784, 'Ipo11': 1764.1816, 'Clasp1': 1440.0, 'Nubpl': 2097.2544, 'Spopl': 1271.3716, 'Rhbdd1': 1372.1356, '2700049A03Rik': 1177.2404, 'Prelid2': 1755.0436, 'Ccdc93': 1384.5172, 'Polq': 1267.9236, 'Ccdc150': 680.0, 'Micu3': 1621.2408, 'Rtf1': 1893.3832, 'Rad54l2': 2183.726, 'Tbl1xr1': 2574.5556, 'Ankrd17': 2170.6428, 'Akap12': 970.6428, 'Bicc1': 1920.004, 'Trps1': 1272.4228, 'Slco2a1': 1033.304, 'Loxl2': 1526.252, 'Spred2': 2579.3636, 'Cnnm2': 2130.556, 'Rnf111': 963.1876, 'Naa15': 2065.1064, 'Trim33': 2360.0, 'Erc1': 2440.0, 'Rasa2': 1724.76, 'Msi2': 734.6088, 'Adgrv1': 2160.0, 'Eri3': 1372.3372, 'Itpr3': 1665.2048, 'Rnf216': 1850.2768, 'Zcchc14': 1047.0092, 'Dusp16': 2324.4628, 'Usp48': 1800.0, 'Parvb': 1840.0, 'Gask1b': 711.5148, 'Ppargc1b': 2605.8476, 'Tram2': 1192.2012, 'Acot7': 1659.9084, 'Opa1': 1053.0652, 'Usp47': 1630.264, 'Efr3a': 1901.2804, 'Hdlbp': 2379.5136, 'Phlpp1': 1531.808, 'Creb1': 1430.582, 'Ube3c': 1483.9148, 'Adgra3': 2270.1752, 'Pot1a': 1080.0, 'Arhgef18': 1165.7608, 'Ano10': 1057.598, 'Fig4': 1823.378, 'Psme4': 2286.252, 'Fbxw11': 3561.64, 'Slf1': 1182.666, 'Rbm26': 2017.6916, 'Adk': 2529.9288, 'Slc38a1': 1796.0888, 'Ncald': 0.0, 'Scaf8': 2080.0, 'Ift140': 624.182, 'Galnt10': 1220.5788, 'Elovl5': 2088.1748, 'Zcchc7': 2003.9508, 'Gldc': 1600.0, 'Mthfd1': 1320.0, 'Ell2': 1937.188, 'Rassf3': 2223.4892, 'Spats2': 1155.008, 'Atxn7': 1773.684, 'Galnt2': 1151.7876, 'Braf': 2874.85, 'Kif18a': 1373.2524, 'Hectd1': 1644.3492, 'Mtss1': 2100.5988, 'Taok1': 1670.24, 'Pcca': 1348.1964, 'Cyrib': 967.9296, 'Uhrf2': 1950.9536, 'Gatb': 600.0, 'Tnrc6a': 2200.0, 'Usp3': 2400.0, 'Trim24': 2280.0, 'Brwd1': 1497.3628, 'Wasf3': 2551.7632, 'Alms1': 1714.2688, 'Rasa1': 2279.9916, 'Stk24': 2382.9548, 'Mettl8': 1303.4464, 'Cstf3': 1073.3116, 'Spg11': 1135.9768, 'Rap1a': 2353.1152, 'Hibadh': 1566.0104, 'Cald1': 203.7764, 'Pdk3': 1715.0656, 'Dhx35': 1527.2672, 'Evc2': 1205.962, 'Fntb': 1654.9152, 'St6gal1': 4498.3364, 'Sumf1': 1603.024, 'Tbc1d12': 1652.8732, 'Cdc73': 1899.12, 'Kdm5a': 1272.4228, 'Fermt2': 2176.0172, 'Atf7': 1703.8516, 'Ubr2': 1333.2516, 'Fam13b': 2061.0992, 'Gigyf2': 1947.5336, 'Ppa2': 1287.5536, 'Jak1': 3762.592, 'Nrd1': 2089.0824, 'Picalm': 2254.2488, 'Cdk17': 2706.9008, 'Fyco1': 1016.616, 'Dicer1': 1511.9692, 'Csnk1g3': 1012.66, 'Ttc27': 1351.8068, 'Usp45': 1205.378, 'Kdm5b': 2267.8904, 'Kcnq3': 544.4396, 'Sec63': 1548.8464, 'Mapre2': 2226.67, 'St7l': 1577.6356, 'E130308A19Rik': 2660.43, 'Cnot1': 1245.1048, 'Lars2': 1308.9272, 'Ago2': 3168.1276, 'Specc1l': 1792.8644, 'Phldb2': 1043.9004, 'Phc3': 1556.6492, 'Wasf2': 2036.72, 'Lrrc20': 1220.9128, 'Slc49a4': 2165.1048, 'Gak': 1716.0888, 'Cdk8': 2007.5536, 'Smc5': 1929.0856, 'Pgbd5': 801.39, 'Zc3h3': 2104.5032, 'Lpgat1': 2517.9252, 'Nup214': 1383.988, 'Sfswap': 933.032, 'Pced1b': 400.0, 'Ecpas': 2294.1108, 'Ldah': 1720.0, 'Tbc1d16': 618.1664, 'Wdr37': 1563.8804, 'Snx30': 1874.356, 'Hace1': 2102.456, 'Hycc2': 1908.9084, 'Zfp609': 1823.5036, 'Vezt': 1504.0316, 'Cnksr3': 2870.5136, 'Zdhhc17': 2083.1004, 'Tmem63c': 85.504, 'Itpk1': 2139.9204, 'Larp4b': 1111.5416, 'Serinc5': 782.4252, 'Lhfpl2': 2498.93, 'Fcho2': 1795.904, 'Arl15': 2192.1116, 'Tdrd3': 2026.6008, 'Marchf6': 2278.5748, 'Gramd4': 1254.026, 'Rbm27': 1552.8204, 'Sh3tc2': 833.9808, 'Mapk4': 1257.7, 'Suco': 1472.0308, 'Epc2': 2903.8084, 'Tlk1': 2458.4756, 'Lgr4': 2429.2068, 'Tmem131l': 2247.5536, 'Patj': 1620.4808, 'Foxj3': 2682.0596, 'Uba6': 1377.1516, 'Fbxw8': 1521.0088, 'Arhgap35': 2222.7148, 'Arhgef10': 420.77, 'L3mbtl3': 2009.6504, 'Lrp12': 2800.0, 'Lnpep': 2354.254, 'Cers6': 3223.8896, 'Cbfa2t2': 3183.7956, 'Dpy19l1': 2171.4472, 'Peak1': 1847.14, 'Shprh': 905.0636, 'Mtmr12': 2001.8164, 'Fbxl4': 1360.5212, 'Ube3a': 1435.8944, 'Relch': 1720.836, 'Chd6': 2999.0168, 'Rgs12': 875.3672, 'Nrip1': 2055.62, 'Nbeal1': 1681.1252, 'Nudcd3': 2739.5864, 'Fnip1': 2768.3088, 'Lin52': 1323.144, 'Ppip5k2': 1595.2756, 'Cenpe': 1062.5856, 'Atp2c1': 2108.9132, 'Atg4c': 886.312, 'Nek1': 441.812, 'Galk2': 1720.0, 'Sash1': 1979.976, 'Utp20': 1350.2888, 'Tmem161b': 2224.028, 'Golim4': 2682.4808, 'Ralgps1': 1955.3468, 'Arid2': 2470.136, 'Rapgef6': 2322.18, 'Myh10': 1813.1984, 'Phactr4': 1417.3628, 'Pds5b': 2461.068, 'Exoc6': 1501.04, 'Ccdc122': 40.0, 'Ankhd1': 1312.9828, 'Rnf169': 1325.2496, 'Sobp': 6960.0, 'Mroh1': 1265.1972, 'Sestd1': 1970.078, 'Tut4': 1729.9796, 'Prtg': 1850.886, 'Intu': 1225.1424, 'Tmem245': 1766.8976, 'Wdfy2': 2458.31, 'Ap4e1': 1327.5536, 'Ccdc88a': 2899.7808, 'Exph5': -80.0, 'Nt5dc1': 2040.0, 'Trappc8': 1986.5232, 'Exoc6b': 2486.44, 'Vps13b': 3219.716, 'Heatr5a': 1185.7764, 'Ubn2': 1522.2412, 'Chd9': 1063.9436, 'Tnpo3': 1640.0, 'Wwp1': 2332.504, 'Hmbox1': 2377.8232, 'Tmcc1': 2287.1356, 'R3hcc1l': 2078.86, 'Usp46': 1349.3544, 'Wdr25': 707.0384, 'Dgkd': 2810.0664, 'Ralgapb': 2261.592, 'Aak1': 1688.1348, 'Ubr3': 1853.7512, 'Frs2': 3208.1888, 'Togaram1': 553.4952, 'Ep300': 1979.2892, 'Zfp236': 1772.2748, 'Ncapd3': 1356.0828, 'Dixdc1': 1830.5752, 'Lcorl': 2640.0, 'Brip1': 1684.146, 'Ylpm1': 1849.6768, 'Commd10': 1629.7608, 'Arap2': 2230.9564, 'Cdc42se2': 2126.1456, 'Klhl2': 1571.4088, 'Uvrag': 2094.832, 'Sfxn5': 1118.972, 'Inpp5f': 1571.568, 'Creb3l2': 1713.834, 'Lpp': 839.1068, '8030462N17Rik': 2177.382, 'Depdc1b': 2405.1048, 'Cep120': 1000.836, 'Rab3gap1': 1929.9476, 'Dpy19l3': 1485.1048, 'Sik2': 2560.0, 'Slc38a9': 1071.4068, 'Fam168a': 2327.5536, 'Rps6kc1': 728.2612, 'Scai': 1260.858, 'Zfp277': 1693.2988, 'Cnot6l': 2403.586, 'Ldlrad3': 1831.6, 'Tyw1': 831.966, 'Gbf1': 2051.2024, 'Plekhh1': 1230.7136, 'Babam2': 2129.2592, 'Pxdn': 1583.988, 'Pik3c3': 1624.0576, 'Syt14': 2103.16, 'Shq1': 1237.706, 'Ppp1r9a': 1116.73, 'R3hdm1': 3249.8504, 'Cdc42bpb': 2145.5892, 'Thada': 920.0, 'Ap1s3': 1603.104, 'Pcmtd1': 2187.4576, 'Igf2bp2': 3167.5056, 'Inpp5a': 2680.8428, 'Ric8b': 1908.6732, 'Usp24': 2038.4464, 'Dennd4c': 1157.316, 'Prr14l': 1346.228, 'Akap9': 1962.6448, 'Zgrf1': 1601.772, 'Tnrc6c': 2707.946, 'Rcor1': 1785.4724, 'Arid4b': 2520.2968, 'Tex2': 648.5032, 'Armh3': 2142.7188, 'Helz': 1106.5064, 'Tbc1d22b': 1427.954, 'B4galnt3': 2086.252, 'Uggt1': 2168.338, 'Znrf2': 2221.1892, 'Cep162': 1120.0, 'Eml4': 2586.4188, 'Mms22l': 1434.7172, 'Gdpd5': 1423.988, 'Ctif': 1061.9632, 'Eipr1': 2654.2972, 'Lyrm4': 1204.6612, 'Map4k5': 2155.7096, 'Ninl': 1723.704, 'Sec24b': 1599.134, 'Osbpl1a': 2079.322, 'Acsl4': 2359.9088, 'Fndc3a': 3072.5624, 'Tsc22d1': 762.0664, 'Hook3': 2154.2648, 'Kif1b': 2305.2452, 'Eya3': 1967.5536, 'Mm2pr': 440.0, 'Wac': 1040.152, 'BC065397': 1094.5556, 'Schip1': 938.2764, '2210408F21Rik': -80.0}\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_excel('gene_list.xlsx') # A list of genes for which transcription elongation rates have been estimated in mESCs by Jonkers et al. in the times spanning  12.5-25 minutes following flavopiridol treatment \n",
    "\n",
    "genedata = dict(zip(df['Gene name'].tolist(), df['Rate (bp/min)'].tolist()))\n",
    "\n",
    "print(genedata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0e64ff5",
   "metadata": {},
   "source": [
    "### Determine gene coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5c8928a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Kdm2a</th>\n",
       "      <td>chr19</td>\n",
       "      <td>4364447</td>\n",
       "      <td>4448313</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ankib1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>3740000</td>\n",
       "      <td>3852925</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pid1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>84014017</td>\n",
       "      <td>84341901</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Wapl</th>\n",
       "      <td>chr14</td>\n",
       "      <td>34395885</td>\n",
       "      <td>34469940</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Kdm2b</th>\n",
       "      <td>chr5</td>\n",
       "      <td>123008728</td>\n",
       "      <td>123127886</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chromosome      start        end strand\n",
       "gene_name                                        \n",
       "Kdm2a          chr19    4364447    4448313      -\n",
       "Ankib1          chr5    3740000    3852925      -\n",
       "Pid1            chr1   84014017   84341901      -\n",
       "Wapl           chr14   34395885   34469940      +\n",
       "Kdm2b           chr5  123008728  123127886      -"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_coordinates = pd.DataFrame({'gene_name': [], 'chromosome': [], 'start': [], 'end': [], 'strand': []})\n",
    "for gene in genedata:\n",
    "    try:\n",
    "        a = gencode_genes.loc[gene]['seqname']\n",
    "        b = gencode_genes.loc[gene]['start']\n",
    "        c = gencode_genes.loc[gene]['end']\n",
    "        d = gencode_genes.loc[gene]['strand']\n",
    "        gene_coordinates.loc[len(gene_coordinates)] = [gene, a, b, c, d]\n",
    "    except Exception as e:\n",
    "        pass\n",
    "gene_coordinates = gene_coordinates.set_index('gene_name')\n",
    "gene_coordinates.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bff528c",
   "metadata": {},
   "source": [
    "### Select exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "69febec2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                   seqname   source feature  \\\n",
      "attribute                                                                     \n",
      "ID=exon:ENSMUST00000162897.2:1;Parent=ENSMUST00...    chr1   HAVANA    exon   \n",
      "ID=exon:ENSMUST00000162897.2:2;Parent=ENSMUST00...    chr1   HAVANA    exon   \n",
      "ID=exon:ENSMUST00000159265.2:1;Parent=ENSMUST00...    chr1   HAVANA    exon   \n",
      "ID=exon:ENSMUST00000159265.2:2;Parent=ENSMUST00...    chr1   HAVANA    exon   \n",
      "ID=exon:ENSMUST00000070533.5:1;Parent=ENSMUST00...    chr1   HAVANA    exon   \n",
      "...                                                    ...      ...     ...   \n",
      "ID=exon:ENSMUST00000084013.1:1;Parent=ENSMUST00...    chrM  ENSEMBL    exon   \n",
      "ID=exon:ENSMUST00000082414.1:1;Parent=ENSMUST00...    chrM  ENSEMBL    exon   \n",
      "ID=exon:ENSMUST00000082418.1:1;Parent=ENSMUST00...    chrM  ENSEMBL    exon   \n",
      "ID=exon:ENSMUST00000082419.1:1;Parent=ENSMUST00...    chrM  ENSEMBL    exon   \n",
      "ID=exon:ENSMUST00000082421.1:1;Parent=ENSMUST00...    chrM  ENSEMBL    exon   \n",
      "\n",
      "                                                      start      end strand  \\\n",
      "attribute                                                                     \n",
      "ID=exon:ENSMUST00000162897.2:1;Parent=ENSMUST00...  3283832  3286567      -   \n",
      "ID=exon:ENSMUST00000162897.2:2;Parent=ENSMUST00...  3276124  3277540      -   \n",
      "ID=exon:ENSMUST00000159265.2:1;Parent=ENSMUST00...  3283662  3285855      -   \n",
      "ID=exon:ENSMUST00000159265.2:2;Parent=ENSMUST00...  3276746  3277540      -   \n",
      "ID=exon:ENSMUST00000070533.5:1;Parent=ENSMUST00...  3740775  3741721      -   \n",
      "...                                                     ...      ...    ...   \n",
      "ID=exon:ENSMUST00000084013.1:1;Parent=ENSMUST00...     9877    10173      +   \n",
      "ID=exon:ENSMUST00000082414.1:1;Parent=ENSMUST00...    10167    11544      +   \n",
      "ID=exon:ENSMUST00000082418.1:1;Parent=ENSMUST00...    11742    13565      +   \n",
      "ID=exon:ENSMUST00000082419.1:1;Parent=ENSMUST00...    13552    14070      -   \n",
      "ID=exon:ENSMUST00000082421.1:1;Parent=ENSMUST00...    14145    15288      +   \n",
      "\n",
      "                                                         gene_type  \n",
      "attribute                                                           \n",
      "ID=exon:ENSMUST00000162897.2:1;Parent=ENSMUST00...  protein_coding  \n",
      "ID=exon:ENSMUST00000162897.2:2;Parent=ENSMUST00...  protein_coding  \n",
      "ID=exon:ENSMUST00000159265.2:1;Parent=ENSMUST00...  protein_coding  \n",
      "ID=exon:ENSMUST00000159265.2:2;Parent=ENSMUST00...  protein_coding  \n",
      "ID=exon:ENSMUST00000070533.5:1;Parent=ENSMUST00...  protein_coding  \n",
      "...                                                            ...  \n",
      "ID=exon:ENSMUST00000084013.1:1;Parent=ENSMUST00...  protein_coding  \n",
      "ID=exon:ENSMUST00000082414.1:1;Parent=ENSMUST00...  protein_coding  \n",
      "ID=exon:ENSMUST00000082418.1:1;Parent=ENSMUST00...  protein_coding  \n",
      "ID=exon:ENSMUST00000082419.1:1;Parent=ENSMUST00...  protein_coding  \n",
      "ID=exon:ENSMUST00000082421.1:1;Parent=ENSMUST00...  protein_coding  \n",
      "\n",
      "[759030 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "gencode_exons = gencode[(gencode.feature == \"exon\")][['seqname', 'source', 'feature','start', 'end', 'strand', 'attribute']].copy().reset_index().drop('index', axis=1)\n",
    "\n",
    "gencode_exons[\"gene_name\"], gencode_exons[\"gene_type\"] = zip(*gencode_exons.attribute.apply(lambda x: gene_info(x)))\n",
    "\n",
    "gencode_exons = gencode_exons[gencode_exons['gene_type'] == 'protein_coding'].reset_index().drop('index', axis=1)\n",
    "\n",
    "gencode_exons = gencode_exons.set_index('gene_name')\n",
    "\n",
    "gencode_exons1 = gencode_exons.set_index('attribute')\n",
    "\n",
    "print(gencode_exons1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c45be51",
   "metadata": {},
   "source": [
    "### Exclude outliers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae4a7335",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = np.array(list(genedata.values()))\n",
    "\n",
    "# Calculate the interquartile range (IQR)\n",
    "q1 = np.percentile(data, 25)\n",
    "q3 = np.percentile(data, 75)\n",
    "iqr = q3 - q1\n",
    "\n",
    "# Define the lower and upper bounds for outliers\n",
    "lower_bound = q1 - 1.5 * iqr\n",
    "upper_bound = q3 + 1.5 * iqr\n",
    "\n",
    "# Exclude the outliers from the data\n",
    "data_excl = data[(data >= lower_bound) & (data <= upper_bound)]\n",
    "\n",
    "genedata1 = {}\n",
    "for gene in genedata:\n",
    "    if genedata[gene] in data_excl:\n",
    "        genedata1[gene] = genedata[gene]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69d28eff",
   "metadata": {},
   "source": [
    "### Count exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33278e44",
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_counts = pd.DataFrame({'gene_name': [], 'exon_count': [], 'Rate (bp/min)':[]})\n",
    "for gene in genedata1:\n",
    "    exon_count = 0\n",
    "\n",
    "    try: \n",
    "        a = gene_coordinates.loc[gene]['chromosome']\n",
    "        b = gene_coordinates.loc[gene]['start']\n",
    "        c = gene_coordinates.loc[gene]['end']\n",
    "        d = gene_coordinates.loc[gene]['strand']\n",
    "        for attribute in gencode_exons.loc[gene]['attribute']:\n",
    "            if \"Ensembl\" in attribute:\n",
    "                start = gencode_exons1.loc[attribute]['start']\n",
    "                end = gencode_exons1.loc[attribute]['end']\n",
    "                if d == '+':\n",
    "                    if end < b+30000:\n",
    "                        if start > b+7500:\n",
    "                            exon_count = exon_count + 1\n",
    "                if d == '-':\n",
    "                    if start > c-30000:\n",
    "                        if end < c-7500:\n",
    "                            exon_count = exon_count + 1\n",
    "\n",
    "    except:\n",
    "        pass\n",
    "    if exon_count >-1:\n",
    "        exon_counts.loc[len(exon_counts)] = [gene, exon_count, genedata[gene]]\n",
    "\n",
    "print(exon_counts) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5562fdf5",
   "metadata": {},
   "source": [
    "### Plot the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "385b47e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in data\n",
    "x = np.array(exon_counts['exon_count'])\n",
    "y = np.array(exon_counts['Rate (bp/min)'])\n",
    "\n",
    "# Calculate the linear regression line and R-squared value\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(x, y)\n",
    "lin_reg = slope * x + intercept\n",
    "r_squared = r_value**2\n",
    "\n",
    "# Set the figure size\n",
    "plt.figure(figsize=(6, 6))  # width=10 inches, height=7 inches\n",
    "\n",
    "# Create a scatter plot\n",
    "plt.scatter(x, y)\n",
    "\n",
    "# Add the linear regression line to the plot\n",
    "plt.plot(x, lin_reg, color='r')\n",
    "\n",
    "r_squared = r_value**2\n",
    "r_squared_percent = r_squared * 100\n",
    "\n",
    "# Add the R-squared value, p-value, and the number of datapoints (n) to the plot\n",
    "plt.annotate(f\"R-squared = {r_squared_percent:.1f}%\", xy=(0.64, 0.95), xycoords='axes fraction', fontsize=12)\n",
    "plt.annotate('p-value = {:.1e}'.format(p_value), xy=(0.64, 0.90), xycoords='axes fraction',fontsize=12)\n",
    "plt.annotate('n = {}'.format(len(x)), xy=(0.64, 0.85), xycoords='axes fraction', fontsize = 12)\n",
    "\n",
    "# Set the x- and y-axis labels\n",
    "plt.xlabel('Exon count', fontsize=15)\n",
    "plt.ylabel('Elongation rate (bp/min)', fontsize=15)\n",
    "\n",
    "plt.xticks(fontsize=11)\n",
    "plt.yticks(fontsize=11)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "066b38eb",
   "metadata": {},
   "source": [
    "### Determine exon coordinates "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5f5c545",
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_counts = pd.DataFrame({'gene_name': [], 'exon_count': [], 'exons': [], 'exons_coordinates': [], 'Rate (bp/min)':[]})\n",
    "for gene in genedata:\n",
    "    exon_count = 0\n",
    "    exons = []\n",
    "    exon_coordinates = []\n",
    "    try: \n",
    "        a = gene_coordinates.loc[gene]['chromosome']\n",
    "        b = gene_coordinates.loc[gene]['start']\n",
    "        c = gene_coordinates.loc[gene]['end']\n",
    "        d = gene_coordinates.loc[gene]['strand']\n",
    "        for attribute in gencode_exons.loc[gene]['attribute']:\n",
    "            if \"Ensembl\" in attribute:\n",
    "                start = gencode_exons1.loc[attribute]['start']\n",
    "                end = gencode_exons1.loc[attribute]['end']\n",
    "                if d == '+':\n",
    "                    if end < b+30000:\n",
    "                        if start > b+7500:\n",
    "                            exon_count = exon_count + 1\n",
    "                            exon_length = end-start\n",
    "                            exons.append(exon_length)\n",
    "                            exon_coordinates.append((a, start, end))\n",
    "                if d == '-':\n",
    "                    if start > c-30000:\n",
    "                        if end < c-7500:\n",
    "                            exon_count = exon_count + 1\n",
    "                            exon_length = end-start\n",
    "                            exons.append(exon_length)\n",
    "                            exon_coordinates.append((a, start, end))\n",
    "    except:\n",
    "        pass\n",
    "    if exon_count >-1:\n",
    "        exon_counts.loc[len(exon_counts)] = [gene, exon_count, exons, exon_coordinates, genedata[gene]]\n",
    "\n",
    "exon_counts = exon_counts.set_index('gene_name')\n",
    "exon_counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3603e04a",
   "metadata": {},
   "source": [
    "### Determine exon density"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "448c8e21",
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_density = pd.DataFrame({'gene_name': [], 'exon_density': [], 'Rate (bp/min)':[]})\n",
    "try:\n",
    "    for gene in genedata1:\n",
    "        total_exon = sum(exon_counts.loc[gene]['exons'])\n",
    "        exon_den = total_exon/22500\n",
    "        if exon_counts.loc[gene]['exon_count'] >=1:\n",
    "            exon_density.loc[len(exon_density)] = [gene, exon_den, genedata[gene]]      \n",
    "except:\n",
    "    pass\n",
    "\n",
    "print(exon_density)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e04739c",
   "metadata": {},
   "source": [
    "### Plot the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c81f252f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in data\n",
    "x = np.array(exon_density['exon_density']) *100\n",
    "y = np.array(exon_density['Rate (bp/min)']) \n",
    "\n",
    "# Calculate the linear regression line and R-squared value\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(x, y)\n",
    "lin_reg = slope * x + intercept\n",
    "r_squared = r_value**2\n",
    "\n",
    "# Set the figure size\n",
    "plt.figure(figsize=(6, 6)) \n",
    "\n",
    "# Create a scatter plot\n",
    "plt.scatter(x, y)\n",
    "\n",
    "# Add the linear regression line to the plot\n",
    "plt.plot(x, lin_reg, color='r')\n",
    "\n",
    "r_squared = r_value**2\n",
    "r_squared_percent = r_squared * 100\n",
    "\n",
    "# Add the R-squared value, p-value, and the number of datapoints (n) to the plot\n",
    "plt.annotate(f\"R-squared = {r_squared_percent:.1f}%\", xy=(0.64, 0.95), xycoords='axes fraction', fontsize=12)\n",
    "plt.annotate('p-value = {:.2e}'.format(p_value), xy=(0.64, 0.90), xycoords='axes fraction', fontsize=12)\n",
    "plt.annotate('n = {}'.format(len(x)), xy=(0.64, 0.85), xycoords='axes fraction', fontsize=12)\n",
    "\n",
    "# Set the x- and y-axis labels\n",
    "plt.xlabel('Exon density (%)', fontsize=15)\n",
    "plt.ylabel('Elongation rate (bp/min)', fontsize=15)\n",
    "\n",
    "plt.xticks(fontsize=11)\n",
    "plt.yticks(fontsize=11)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfccc4fd",
   "metadata": {},
   "source": [
    "### Determine the GC-content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "649c52e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "GC_content = pd.DataFrame({'Gene_name': [], 'GC_content': [], 'Rate (bp/min)' : []})\n",
    "\n",
    "for gene in genedata1:\n",
    "\n",
    "    try:       \n",
    "        a = gencode_genes.loc[gene]['seqname']\n",
    "        b = gencode_genes.loc[gene]['start']\n",
    "        c = gencode_genes.loc[gene]['end']\n",
    "        d = gencode_genes.loc[gene]['strand']\n",
    "        x = genomepy.Genome(\"GRCm39\").get_seq(a[3:],b,c)\n",
    "        x = (str(x)).upper()\n",
    "        if d == '+':\n",
    "            x = x[7499:30000]\n",
    "        if d == '-':\n",
    "            x = x[-30000:-7499]\n",
    "        gc_count = x.count(\"G\") + x.count(\"C\")\n",
    "        gc_content = gc_count / len(x) * 100\n",
    "        GC_content.loc[len(GC_content)] = [gene, gc_content, genedata[gene]]\n",
    "    except Exception as e:\n",
    "        pass\n",
    "    \n",
    "print(GC_content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1549b008",
   "metadata": {},
   "source": [
    "### Plot the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2761f685",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in data \n",
    "x = np.array(GC_content['GC_content'])\n",
    "y = np.array(GC_content['Rate (bp/min)'])\n",
    "\n",
    "# Calculate the linear regression line and R-squared value\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(x, y)\n",
    "lin_reg = slope * x + intercept\n",
    "r_squared = r_value**2\n",
    "\n",
    "# Set the figure size\n",
    "plt.figure(figsize=(6, 6))\n",
    "\n",
    "# Create a scatter plot\n",
    "plt.scatter(x, y)\n",
    "\n",
    "# Add the linear regression line to the plot\n",
    "plt.plot(x, lin_reg, color='r')\n",
    "\n",
    "r_squared = r_value**2\n",
    "r_squared_percent = r_squared * 100\n",
    "\n",
    "# Add the R-squared value, p-value, and the number of datapoints (n) to the plot\n",
    "plt.annotate(f\"R² = {r_squared_percent:.2f}%\", xy=(0.75, 0.95), xycoords='axes fraction', fontsize=12)\n",
    "plt.annotate('p = {:.2e}'.format(p_value), xy=(0.75, 0.90), xycoords='axes fraction', fontsize = 12)\n",
    "plt.annotate('n = {}'.format(len(x)), xy=(0.75, 0.85), xycoords='axes fraction', fontsize = 12)\n",
    "\n",
    "# Set the x- and y-axis labels\n",
    "plt.xlabel('Guanine-cytosine content (%)', fontsize=15)\n",
    "plt.ylabel('Elongation rate (bp/min)', fontsize=15)\n",
    "\n",
    "plt.xticks(fontsize=11)\n",
    "plt.yticks(fontsize=11)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f1786c5",
   "metadata": {},
   "source": [
    "### Count CpG sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ef5e8fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "CpG_sites = pd.DataFrame({'Gene_name': [], 'CpG_count': [], 'Rate (bp/min)' : []})\n",
    "\n",
    "for gene in genedata1:\n",
    "\n",
    "    try:       \n",
    "        a = gencode_genes.loc[gene]['seqname']\n",
    "        b = gencode_genes.loc[gene]['start']\n",
    "        c = gencode_genes.loc[gene]['end']\n",
    "        d = gencode_genes.loc[gene]['strand']\n",
    "        x = genomepy.Genome(\"GRCm39\").get_seq(a[3:],b,c)\n",
    "        x = (str(x)).upper()\n",
    "        if d == '+':\n",
    "            x = x[7499:30000]\n",
    "        if d == '-':\n",
    "            x = x[-30000:-7499]\n",
    "        CpG_count = x.count(\"CG\")\n",
    "        CpG_sites.loc[len(CpG_sites)] = [gene, CpG_count, genedata[gene]]\n",
    "    except Exception as e:\n",
    "        pass\n",
    "    \n",
    "print(CpG_sites)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62426ac4",
   "metadata": {},
   "source": [
    "### Plot the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afbae1ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in data\n",
    "x = np.array(CpG_sites['CpG_count'])\n",
    "y = np.array(CpG_sites['Rate (bp/min)'])\n",
    "\n",
    "# Calculate the linear regression line and R-squared value\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(x, y)\n",
    "lin_reg = slope * x + intercept\n",
    "r_squared = r_value**2\n",
    "\n",
    "# Set the figure size\n",
    "plt.figure(figsize=(6, 6))\n",
    "\n",
    "# Create a scatter plot\n",
    "plt.scatter(x, y)\n",
    "\n",
    "# Add the linear regression line to the plot\n",
    "plt.plot(x, lin_reg, color='r')\n",
    "\n",
    "r_squared = r_value**2\n",
    "r_squared_percent = r_squared * 100\n",
    "\n",
    "# Add the R-squared value, p-value, and the number of datapoints (n) to the plot\n",
    "plt.annotate(f\"R² = {r_squared_percent:.2f}%\", xy=(0.75, 0.95), xycoords='axes fraction', fontsize=12)\n",
    "plt.annotate('p = {:.2e}'.format(p_value), xy=(0.75, 0.90), xycoords='axes fraction', fontsize = 12)\n",
    "plt.annotate('n = {}'.format(len(x)), xy=(0.75, 0.85), xycoords='axes fraction', fontsize = 12)\n",
    "\n",
    "plt.xticks(fontsize=11)\n",
    "plt.yticks(fontsize=11)\n",
    "\n",
    "# Set the x- and y-axis labels\n",
    "plt.xlabel('CpG sites count', fontsize=15)\n",
    "plt.ylabel('Elongation rate (bp/min)', fontsize=15)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
